### ISSUE TITLE
FreeBSD: failed building after commit a96339b

### ISSUE THREAD

**Issue by alexax66** [SerenityOS/serenity#17713]
Failed build after commit a96339b

```
FAILED: Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o 
/usr/bin/clang++ -DLibCore_EXPORTS -I/home/alexax/src/serenity/Meta/Lagom/../.. -I/home/alexax/src/serenity/Meta/Lagom/../../Userland -I/home/alexax/src/serenity/Meta/Lagom/../../Userland/Libraries -I/home/alexax/src/serenity/Meta/Lagom/../../Userland/Services -I/home/alexax/src/serenity/Build/lagom -I/home/alexax/src/serenity/Build/lagom/Userland/Libraries -I/home/alexax/src/serenity/Build/lagom/Userland/Services -fPIC -Wall -Wextra -Werror -Wno-implicit-const-int-float-conversion -Wno-literal-suffix -Wno-maybe-uninitialized -Wno-unknown-warning-option -Wno-unused-command-line-argument -fsigned-char -fno-exceptions -fdiagnostics-color=always -g1 -O2 -fno-semantic-interposition -fPIC -Wno-overloaded-virtual -Wno-user-defined-literals -fconstexpr-steps=16777216 -Wno-unused-private-field -std=c++20 -MD -MT Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o -MF Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o.d -o Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o -c /home/alexax/src/serenity/Userland/Libraries/LibCore/Socket.cpp
/home/alexax/src/serenity/Userland/Libraries/LibCore/Socket.cpp:365:19: error: variable has incomplete type 'struct xucred'
    struct xucred creds = {};
                  ^
/home/alexax/src/serenity/Userland/Libraries/LibCore/Socket.cpp:365:12: note: forward declaration of 'xucred'
    struct xucred creds = {};
           ^
1 error generated.
```

**Comment by ADKaster** [comment]
Can you check latest? Some support was added for a SunOS derivative that I think might have affected the include order on BSD platforms

**Comment by alexax66** [comment]
Unsuccessfully:

```
[26/133] Building CXX object Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o
FAILED: Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o 
/usr/bin/clang++ -DLibCore_EXPORTS -I/home/alexax/src/serenity/Meta/Lagom/../.. -I/home/alexax/src/serenity/Meta/Lagom/../../Userland -I/home/alexax/src/serenity/Meta/Lagom/../../Userland/Libraries -I/home/alexax/src/serenity/Meta/Lagom/../../Userland/Services -I/home/alexax/src/serenity/Build/lagom -I/home/alexax/src/serenity/Build/lagom/Userland/Libraries -I/home/alexax/src/serenity/Build/lagom/Userland/Services -fPIC -Wall -Wextra -Werror -Wno-implicit-const-int-float-conversion -Wno-literal-suffix -Wno-maybe-uninitialized -Wno-unknown-warning-option -Wno-unused-command-line-argument -fsigned-char -fno-exceptions -fdiagnostics-color=always -g1 -O2 -fno-semantic-interposition -fPIC -Wno-overloaded-virtual -Wno-user-defined-literals -fconstexpr-steps=16777216 -Wno-unused-private-field -std=c++20 -MD -MT Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o -MF Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o.d -o Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o -c /home/alexax/src/serenity/Userland/Libraries/LibCore/Socket.cpp
/home/alexax/src/serenity/Userland/Libraries/LibCore/Socket.cpp:365:19: error: variable has incomplete type 'struct xucred'
    struct xucred creds = {};
                  ^
/home/alexax/src/serenity/Userland/Libraries/LibCore/Socket.cpp:365:12: note: forward declaration of 'xucred'
    struct xucred creds = {};
           ^
1 error generated.
[29/133] Building CXX object Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/System.cpp.o
ninja: build stopped: subcommand failed.
FAILED: lagom-prefix/src/lagom-stamp/lagom-build /home/alexax/src/serenity/Build/superbuild-x86_64/lagom-prefix/src/lagom-stamp/lagom-build 
cd /home/alexax/src/serenity/Build/lagom && /zdata/home/alexax/src/serenity/Toolchain/Local/cmake/bin/cmake --build .
ninja: build stopped: subcommand failed.
```

**Comment by ADKaster** [comment]
This looks like we're just missing an 
```
#ifdef AK_OS_BSD_GENERIC
#    include <sys/ucred.h>
#endif
```
from the top of LibCore/Socket.cpp?

Or is sys/ucred.h a FreeBSD only thing, that the other BSDs don't share?

**Comment by alexax66** [comment]
I tried it:

```
diff --git a/Userland/Libraries/LibCore/Socket.cpp b/Userland/Libraries/LibCore/Socket.cpp
index dfc400322f..12388731f9 100644
--- a/Userland/Libraries/LibCore/Socket.cpp
+++ b/Userland/Libraries/LibCore/Socket.cpp
@@ -5,6 +5,10 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#ifdef AK_OS_BSD_GENERIC
+#    include <sys/ucred.h>
+#endif
+
 #include <LibCore/Socket.h>
 #include <LibCore/System.h>
```

The same error:

```
alexax@ghostbsd-pc ~/s/serenity (master)> ./Meta/serenity.sh image
[0/10] Performing build step for 'lagom'
[26/131] Building CXX object Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o
FAILED: Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o 
/usr/bin/clang++ -DLibCore_EXPORTS -I/home/alexax/src/serenity/Meta/Lagom/../.. -I/home/alexax/src/serenity/Meta/Lagom/../../Userland -I/home/alexax/src/serenity/Meta/Lagom/../../Userland/Libraries -I/home/alexax/src/serenity/Meta/Lagom/../../Userland/Services -I/home/alexax/src/serenity/Build/lagom -I/home/alexax/src/serenity/Build/lagom/Userland/Libraries -I/home/alexax/src/serenity/Build/lagom/Userland/Services -fPIC -Wall -Wextra -Werror -Wno-implicit-const-int-float-conversion -Wno-literal-suffix -Wno-maybe-uninitialized -Wno-unknown-warning-option -Wno-unused-command-line-argument -fsigned-char -fno-exceptions -fdiagnostics-color=always -g1 -O2 -fno-semantic-interposition -fPIC -Wno-overloaded-virtual -Wno-user-defined-literals -fconstexpr-steps=16777216 -Wno-unused-private-field -std=c++20 -MD -MT Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o -MF Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o.d -o Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/Socket.cpp.o -c /home/alexax/src/serenity/Userland/Libraries/LibCore/Socket.cpp
/home/alexax/src/serenity/Userland/Libraries/LibCore/Socket.cpp:369:19: error: variable has incomplete type 'struct xucred'
    struct xucred creds = {};
                  ^
/home/alexax/src/serenity/Userland/Libraries/LibCore/Socket.cpp:369:12: note: forward declaration of 'xucred'
    struct xucred creds = {};
           ^
1 error generated.
[29/131] Building CXX object Userland/Libraries/LibCore/CMakeFiles/LibCore.dir/System.cpp.o
ninja: build stopped: subcommand failed.
FAILED: lagom-prefix/src/lagom-stamp/lagom-build /home/alexax/src/serenity/Build/superbuild-x86_64/lagom-prefix/src/lagom-stamp/lagom-build 
cd /home/alexax/src/serenity/Build/lagom && /zdata/home/alexax/src/serenity/Toolchain/Local/cmake/bin/cmake --build .
ninja: build stopped: subcommand failed.
```

Yes, FreeBSD, NetBSD, OpenBSD have ucred.h in /usr/include/sys

P.S. After some recent commits my hack no longer works

**Comment by ADKaster** [comment]
That error doesn't make much sense. when `__serenity__` is defined by the cross-compiler, we shouldn't pick up a host definition that leads it to think that we're on a BSD system.

Adding the `<sys/ucred.h>` include worked for building LibCore as part of Lagom for me in a FreeBSD VM with `clang++-devel`, so we can make that change at least.

**Comment by nipos** [comment]
Just found this one by accident.
Should have been fixed months ago by #18506
Please re-test and let me know about the result.
